src/rgw/rgw_gc.cc

```diff
@@ -38,17 +37,29 @@ void RGWGC::initialize(CephContext *_cct, RGWRados *_store) {
     char buf[32];
     snprintf(buf, 32, ".%d", i);
     obj_names[i].append(buf);
-  }
 
-  // Create GC queues
-  for (int i = 0; i < max_objs; i++) {
+    std::list<cls_rgw_gc_obj_info> entries;
+    string marker, next_marker;
+    bool truncated = false;
+    auto ret = cls_rgw_gc_list(store->gc_pool_ctx, obj_names[i], marker, 1,
```

Comment:
@cbodley , @mattbenjamin : We had decided to maintain an xattr which will tell us whether or not transition from omap to queue has finished, but I thought that reading non-expired entries from omap will amount to the same thing, hence I have used cls_rgw_gc_list to populate the cache of transitioned objects.
Also in case of absence of a xattr denoting the transition, we had decided to take it as a non transitioned shard - an empty shard without the xattr would also amount it being set as non-transitioned (which can be taken care of in the RGW:process but will need an extra step)

Reply:
does this have to happen during initialization? the alternative would be to initialize the all cache entries to false, and rely on RGWGC::list()/process() to set this flag when they detect empty omap?

Reply:
Is there a possibility that defer_gc() can be called before process() or list()? the logic depends on the cache entries.

Reply:
> Is there a possibility that defer_gc() can be called before process() or list()? the logic depends on the cache entries.

yes, but i think it's okay for defer to see transitioned=false before process() or list() run

> Also , shouldn't the cache be locked across different methods in RGWGC? It is updated in process() and read in defer_gc(), list() and process()

yes, agreed! i think its enough to make these bools atomic